File: RC4.c
#include <stdio.h>
#include <string.h>
void display(int disp[], int length) {
	int l;
    char convert[length];
    for (l = 0; l < length; l++) {
        convert[l] = (char)disp[l];
        printf("%c", convert[l]);
    }
}
int main() {
    int temp = 0, i, l;
    char ptext[256];
    char key[256];
    int s[256];
    int k[256];
    printf("\nENTER PLAIN TEXT: ");
    fgets(ptext, sizeof(ptext), stdin);
    printf("ENTER KEY TEXT: ");
    fgets(key, sizeof(key), stdin);
    int ptextLen = strlen(ptext) - 1;
    int keyLen = strlen(key) - 1;
    int cipher[ptextLen];
    int decrypt[ptextLen];
    int ptexti[ptextLen];
    int keyi[keyLen];
    for (i = 0; i < ptextLen; i++) {
        ptexti[i] = (int)ptext[i];
    }
    for (i = 0; i < keyLen; i++) {
        keyi[i] = (int)key[i];
    }
    for (i = 0; i < 255; i++) {
        s[i] = i;
        k[i] = keyi[i % keyLen];
    }
    int j = 0;
    for (i = 0; i < 255; i++) {
        j = (j + s[i] + k[i]) % 256;
        temp = s[i];
        s[i] = s[j];
        s[j] = temp;
    }
    i = 0;
    j = 0;
    int z = 0;
    for (l = 0; l < ptextLen; l++) {
        i = (l + 1) % 256;
        j = (j + s[i]) % 256;
        temp = s[i];
        s[i] = s[j];
        s[j] = temp;
        z = s[(s[i] + s[j]) % 256];
        cipher[l] = z ^ ptexti[l];
        decrypt[l] = z ^ cipher[l];
    }
    printf("ENCRYPTED: ");
    display(cipher, ptextLen);
    printf("\nDECRYPTED: ");
    display(decrypt, ptextLen);
    return 0;
}

File: rc4.py
def display(disp):
    convert = [chr(char) for char in disp]
    print("".join(convert))

def main():
    temp = 0
    ptext = input("\nEnter Plain Text: ")
    key = input("Enter Key Text: ")
    ptextLen = len(ptext)
    keyLen = len(key)
    cipher = [0] * ptextLen
    decrypt = [0] * ptextLen
    ptexti = [ord(char) for char in ptext]
    keyi = [ord(char) for char in key]
    s = list(range(256))
    k = keyi * (255 // keyLen + 1)
    j = 0

    for i in range(256):
        j = (j + s[i] + k[i]) % 256
        temp = s[i]
        s[i] = s[j]
        s[j] = temp

    i = 0
    j = 0
    z = 0

    for l in range(ptextLen):
        i = (l + 1) % 256
        j = (j + s[i]) % 256
        temp = s[i]
        s[i] = s[j]
        s[j] = temp
        z = s[(s[i] + s[j]) % 256]
        cipher[l] = z ^ ptexti[l]
        decrypt[l] = z ^ cipher[l]

    print("ENCRYPTED: ", end="")
    display(cipher)
    print("\nDECRYPTED: ", end="")
    display(decrypt)

if __name__ == "__main__":
    main()
